enterprise/internal/database/perms_store.go

```diff
@@ -2155,17 +2171,197 @@ FROM
 		AND up.object_ids_ints @> INTSET (ar.id)
 `
 
-type UserPermission struct {
-	Repo      *types.Repo
+var defaultPageSize = 100
+
+var defaultPaginationArgs = database.PaginationArgs{
+	First:   &defaultPageSize,
+	OrderBy: database.OrderBy{{Field: "users.id"}},
+}
+
+type ListRepoPermissionsArgs struct {
+	Query          string
+	PaginationArgs *database.PaginationArgs
+}
+
+type RepoPermission struct {
+	User      *types.User
 	Reason    UserRepoPermissionReason
 	UpdatedAt time.Time
 }
 
+// ListRepoPermissions gets the list of users who has access to the repository, along with the reason
+// and timestamp for each permission.
+func (s *permsStore) ListRepoPermissions(ctx context.Context, repoID api.RepoID, args *ListRepoPermissionsArgs) ([]*RepoPermission, error) {
+	authzParams, err := database.GetAuthzQueryParameters(context.Background(), s.ossDB)
+	if err != nil {
+		return nil, err
+	}
+
+	permsQueryConditions := []*sqlf.Query{}
+	unrestricted := false
+
+	if authzParams.BypassAuthzReasons.NoAuthzProvider {
+		// return all users as auth is bypassed for everyone
+		permsQueryConditions = append(permsQueryConditions, sqlf.Sprintf("TRUE"))
+		unrestricted = true
+	} else {
+		// find if the repo is unrestricted
+		isRepoUnrestricted, err := s.isRepoUnrestricted(ctx, repoID, authzParams)
+		if err != nil {
+			return nil, err
+		}
+
+		unrestricted = isRepoUnrestricted != nil && *isRepoUnrestricted
+
+		if unrestricted {
+			// return all users as repo is unrestricted
+			permsQueryConditions = append(permsQueryConditions, sqlf.Sprintf("TRUE"))
+		} else {
+			if !authzParams.AuthzEnforceForSiteAdmins {
+				// include all site admins
+				permsQueryConditions = append(permsQueryConditions, sqlf.Sprintf("users.site_admin"))
+			}
+
+			// TODO(milan): need to update read path for new data model
+			// load user who has access via permissions syncing
+			permsQueryConditions = append(permsQueryConditions, sqlf.Sprintf(`
+				(
+					SELECT object_ids_ints @> INTSET(repo.id)
+					FROM user_permissions
+					WHERE
+						user_id = users.id
+				)
```

Comment:
This seems slow to me. We should leverage the list of user_ids from the `repo_permissions` table instead. So something similar to this query:
```suggestion
				(
					users.id = ANY(repo_permissions.user_ids_ints)
				)
```

Reply:
This will be a good improvement. 

And if we make join, then the `permissions_updated_at` can be calculated directly without another nested query.